subprojects/gst-plugins-bad/sys/msdk/gstmsdkcaps.c

```diff
@@ -0,0 +1,879 @@
+      if (out->mfx.FrameInfo.Width != 0) {
+        if (func (session, out, out)) {
+          *max_w = out->mfx.FrameInfo.Width;
+          return TRUE;
+        }
+      }
+    }
+
+    w -= width->Step;
+  }
+
+  return FALSE;
+}
+
+static gboolean
+_get_max_height (MsdkSession * session, mfxVideoParam * in,
+    mfxVideoParam * out, IsParamSupportedFunc func,
+    const mfxRange32U * height, guint * max_h)
+{
+  g_return_val_if_fail (func != NULL, FALSE);
+
+  in->mfx.FrameInfo.Width = default_width;
+  in->mfx.FrameInfo.CropW = default_width;
+  out->mfx.FrameInfo.Width = in->mfx.FrameInfo.Width;
+  out->mfx.FrameInfo.CropW = in->mfx.FrameInfo.CropW;
+
+  for (guint h = height->Max; h > *max_h;) {
+    in->mfx.FrameInfo.Height = h;
+    in->mfx.FrameInfo.CropH = h;
+    out->mfx.FrameInfo.Height = in->mfx.FrameInfo.Height;
+    out->mfx.FrameInfo.CropH = in->mfx.FrameInfo.CropH;
+
+    if (func (session, in, out)) {
+      *max_h = in->mfx.FrameInfo.Height;
+      return TRUE;
+    } else {
+      if (out->mfx.FrameInfo.Height != 0) {
+        if (func (session, out, out)) {
+          *max_h = out->mfx.FrameInfo.Height;
+          return TRUE;
+        }
+      }
+    }
+
+    h -= height->Step;
+  }
+
+  return FALSE;
+}
+
+static gboolean
+_get_supported_formats (MsdkSession * session,
+    IsParamSupportedFunc func, guint codec_id,
+    mfxU16 pattern, GValue * supported_fmts)
+{
+  GstVideoFormat fmt;
+  const GValue *gfmt;
+  const gchar *fmt_str = NULL;
+  GValue fmts = G_VALUE_INIT;
+  guint size;
+  mfxVideoParam in, out;
+
+  g_return_val_if_fail (func != NULL, FALSE);
+  g_return_val_if_fail (supported_fmts != NULL, FALSE);
+
+  _init_param (&in, codec_id, pattern);
+  in.mfx.CodecProfile = MFX_PROFILE_UNKNOWN;
+  out = in;
+
+  g_value_init (&fmts, GST_TYPE_LIST);
+  gst_msdk_get_video_format_list (&fmts);
+
+  size = gst_value_list_get_size (&fmts);
+  for (guint i = 0; i < size; i++) {
+    gfmt = gst_value_list_get_value (&fmts, i);
+    fmt_str = g_value_get_string (gfmt);
+    fmt = gst_video_format_from_string (fmt_str);
+
+    if (!_fill_frameinfo (fmt, &in.mfx.FrameInfo))
+      continue;
+
+    out.mfx.FrameInfo.BitDepthLuma = in.mfx.FrameInfo.BitDepthLuma;
+    out.mfx.FrameInfo.BitDepthChroma = in.mfx.FrameInfo.BitDepthChroma;
+    out.mfx.FrameInfo.Shift = in.mfx.FrameInfo.Shift;
+
+    in.mfx.LowPower = MFX_CODINGOPTION_UNKNOWN;
+    if (!func (session, &in, &out)) {
+      in.mfx.LowPower = (out.mfx.LowPower == MFX_CODINGOPTION_ON) ?
+          MFX_CODINGOPTION_OFF : MFX_CODINGOPTION_ON;
+
+      if (!func (session, &in, &out))
+        continue;
+    }
+
+    gst_value_list_append_value (supported_fmts, gfmt);
+  }
+  g_value_unset (&fmts);
+
+  return gst_value_list_get_size (supported_fmts) == 0 ? FALSE : TRUE;
+}
+
+static gboolean
+_enc_is_param_supported (MsdkSession * session,
+    mfxVideoParam * in, mfxVideoParam * out)
+{
+  mfxStatus status = MFXVideoENCODE_Query (session->session, in, out);
```

Comment:
I think there is no problem to invoke MFXVideoENCODE_Query to query encoder capability, but it may not be enough, because when I try "gst-inspect-msdkh264enc", I see more formats such as YUY2, VUYA which are actually not supported for avc encoding. According to [msdk manual](https://github.com/Intel-Media-SDK/MediaSDK/blob/master/doc/mediasdk-man.md#mfxvideoencode_query), attaching mfxExtEncoderCapability may help.

Reply:
Thank you, @mengker33 
I mainly compare the formats with the previous static caps, which also contain some formats that are not in the manual (format: { (string)NV12, (string)I420, (string)YV12, (string)YUY2, (string)UYVY, (string)BGRA }). I will check further and try the mfxExtEncoderCapability structure.

Reply:
@mengker33

[mfxExtEncoderCapability](https://github.com/Intel-Media-SDK/MediaSDK/blob/master/doc/mediasdk-man.md#mfxextencodercapability) looks like a dedicated structure, which is only used to obtain the parameter of MBPerSec, not a general structure.

Just like the usage of the [ext_vsi (mfxExtVideoSignalInfo)](https://github.com/intel-media-ci/gstreamer/blob/921c93a9af48366c49c52a39157b104a2cd3b03f/subprojects/gst-plugins-bad/sys/msdk/gstmsdkenc.c#LL776-L794C71) in msdkenc, BufferID and BufferSz need to be specified, and the parameters supported by [mfxExtVideoSignalInfo ](https://github.com/Intel-Media-SDK/MediaSDK/blob/master/doc/mediasdk-man.md#mfxextvideosignalinfo) are fixed:
```
typedef struct {
    mfxExtBuffer    Header;
    mfxU16        VideoFormat;
    mfxU16        VideoFullRange;
    mfxU16        ColourDescriptionPresent;
    mfxU16        ColourPrimaries;
    mfxU16        TransferCharacteristics;
    mfxU16        MatrixCoefficients;
} mfxExtVideoSignalInfo;
```

Also for the [mfxExtEncoderCapability](https://github.com/Intel-Media-SDK/MediaSDK/blob/master/doc/mediasdk-man.md#mfxextencodercapability) structure, only one parameter is supported and that is MBPerSec:
```
typedef struct {
    mfxExtBuffer Header;

    mfxU32      MBPerSec;
    mfxU16      reserved[58];
} mfxExtEncoderCapability;
```

There is also such an implementation in mediaSDK [Query mode 4: Query should do a single thing - report MB processing rate](https://github.com/Intel-Media-SDK/MediaSDK/blob/c98c06a427422bac7b6300e685de10aacbc2141a/_studio/mfx_lib/encode_hw/h264/src/mfx_h264_encode_hw.cpp#LL508-L540C6).

Reply:
@mengker33 

Which path are you running, MSDK or VPL?

If it is MSDK path, the output of "gst-inspect-1.0 msdkh264enc" should be the same as the static caps, the MSDK path does not support obtaining dynamic caps, and the structures and APIs required to obtain dynamic caps are not supported in the MSDK path:
```
Pad Templates:
  SINK template: 'sink'
    Availability: Always
    Capabilities:
      video/x-raw
                 format: { (string)NV12, (string)I420, (string)YV12, (string)YUY2, (string)UYVY, (string)BGRA }
                  width: [ 1, 2147483647 ]
                 height: [ 1, 2147483647 ]
              framerate: [ 0/1, 2147483647/1 ]
         interlace-mode: progressive
      video/x-raw(memory:DMABuf)
                 format: NV12
                  width: [ 1, 2147483647 ]
                 height: [ 1, 2147483647 ]
              framerate: [ 0/1, 2147483647/1 ]
         interlace-mode: progressive
      video/x-raw(memory:VAMemory)
                 format: NV12
                  width: [ 1, 2147483647 ]
                 height: [ 1, 2147483647 ]
              framerate: [ 0/1, 2147483647/1 ]
         interlace-mode: progressive
  
  SRC template: 'src'
    Availability: Always
    Capabilities:
      video/x-h264
                profile: { (string)main, (string)baseline, (string)extended, (string)high, (string)constrained-baseline, (string)constrained-high, (string)progressive-high }
                  width: [ 1, 2147483647 ]
                 height: [ 1, 2147483647 ]
              framerate: [ 0/1, 2147483647/1 ]
              alignment: au
          stream-format: byte-stream
```

If it is VPL path, the output of "gst-inspect-1.0 msdkh264enc" only has NV12 format:
```
Pad Templates:
  SINK template: 'sink'
    Availability: Always
    Capabilities:
      video/x-raw
                 format: { (string)NV12 }
                  width: [ 32, 4096 ]
                 height: [ 32, 4096 ]
         interlace-mode: progressive
      video/x-raw(memory:DMABuf)
                 format: { (string)NV12 }
                  width: [ 32, 4096 ]
                 height: [ 32, 4096 ]
         interlace-mode: progressive
      video/x-raw(memory:VAMemory)
                 format: NV12
                  width: [ 32, 4096 ]
                 height: [ 32, 4096 ]
         interlace-mode: progressive
  
  SRC template: 'src'
    Availability: Always
    Capabilities:
      video/x-h264
                profile: { (string)baseline, (string)constrained-baseline, (string)main, (string)high, (string)constrained-high, (string)progressive-high }
                  width: [ 32, 4096 ]
                 height: [ 32, 4096 ]
              alignment: au
          stream-format: byte-stream
```

Reply:
> @mengker33
> 
> Which path are you running, MSDK or VPL?
> 
> If it is MSDK path, the output of "gst-inspect-1.0 msdkh264enc" should be the same as the static caps, the MSDK path does not support obtaining dynamic caps, and the structures and APIs required to obtain dynamic caps are not supported in the MSDK path:
> 
> ```
> Pad Templates:
>   SINK template: 'sink'
>     Availability: Always
>     Capabilities:
>       video/x-raw
>                  format: { (string)NV12, (string)I420, (string)YV12, (string)YUY2, (string)UYVY, (string)BGRA }
>                   width: [ 1, 2147483647 ]
>                  height: [ 1, 2147483647 ]
>               framerate: [ 0/1, 2147483647/1 ]
>          interlace-mode: progressive
>       video/x-raw(memory:DMABuf)
>                  format: NV12
>                   width: [ 1, 2147483647 ]
>                  height: [ 1, 2147483647 ]
>               framerate: [ 0/1, 2147483647/1 ]
>          interlace-mode: progressive
>       video/x-raw(memory:VAMemory)
>                  format: NV12
>                   width: [ 1, 2147483647 ]
>                  height: [ 1, 2147483647 ]
>               framerate: [ 0/1, 2147483647/1 ]
>          interlace-mode: progressive
>   
>   SRC template: 'src'
>     Availability: Always
>     Capabilities:
>       video/x-h264
>                 profile: { (string)main, (string)baseline, (string)extended, (string)high, (string)constrained-baseline, (string)constrained-high, (string)progressive-high }
>                   width: [ 1, 2147483647 ]
>                  height: [ 1, 2147483647 ]
>               framerate: [ 0/1, 2147483647/1 ]
>               alignment: au
>           stream-format: byte-stream
> ```
> 
> If it is VPL path, the output of "gst-inspect-1.0 msdkh264enc" only has NV12 format:
> 
> ```
> Pad Templates:
>   SINK template: 'sink'
>     Availability: Always
>     Capabilities:
>       video/x-raw
>                  format: { (string)NV12 }
>                   width: [ 32, 4096 ]
>                  height: [ 32, 4096 ]
>          interlace-mode: progressive
>       video/x-raw(memory:DMABuf)
>                  format: { (string)NV12 }
>                   width: [ 32, 4096 ]
>                  height: [ 32, 4096 ]
>          interlace-mode: progressive
>       video/x-raw(memory:VAMemory)
>                  format: NV12
>                   width: [ 32, 4096 ]
>                  height: [ 32, 4096 ]
>          interlace-mode: progressive
>   
>   SRC template: 'src'
>     Availability: Always
>     Capabilities:
>       video/x-h264
>                 profile: { (string)baseline, (string)constrained-baseline, (string)main, (string)high, (string)constrained-high, (string)progressive-high }
>                   width: [ 32, 4096 ]
>                  height: [ 32, 4096 ]
>               alignment: au
>           stream-format: byte-stream
> ```

I use vpl path to build gstreamer. Is there any update of this patch? I still get other formats instead of  just 'NV12' when gst-inspect-1.0 msdkh264enc on my local TGL.
Pad Templates:
  SINK template: 'sink'
    Availability: Always
    Capabilities:
      video/x-raw
                 format: { (string)NV12, (string)YUY2, (string)BGRA, (string)BGRx, (string)VUYA }
                  width: [ 32, 4096 ]
                 height: [ 32, 4096 ]
         interlace-mode: progressive
      video/x-raw(memory:DMABuf)
                 format: { (string)NV12, (string)YUY2, (string)BGRA, (string)BGRx, (string)VUYA }
                  width: [ 32, 4096 ]
                 height: [ 32, 4096 ]
         interlace-mode: progressive
      video/x-raw(memory:VAMemory)
                 format: NV12
                  width: [ 32, 4096 ]
                 height: [ 32, 4096 ]
         interlace-mode: progressive

Reply:
The latest patches are submitted [here](https://github.com/intel-media-ci/gstreamer/pull/51), I didn't find the function to reopen this PR, basically there is no update. You can try the new patches. In addition, I have updated oneVPL to the latest version. I am not sure if the driver will affect this.

Reply:
@mengker33 is it because your code is relatively early, and there are still many bugs that have not been fixed?

Reply:
Sorry, I didn't apply your latest patch. Now gst-inspect-1.0 gives the correct caps.